internal/dbt/service.go

```diff
@@ -0,0 +1,413 @@
+	if override != "" {
+		clientOptions = append(clientOptions, sdk.WithUrlOverride(override))
+	}
+
+	client := sdk.NewClient(ctx, config.Domain, config.ApiUser, config.ApiSecret, clientOptions...)
+
+	return &DbtService{
+		dataSourceId:         config.DataSourceId,
+		accessProviderClient: client.AccessProvider(),
+		logger:               config.BaseLogger,
+	}
+}
+
+func (s *DbtService) RunDbt(ctx context.Context, dbtFile string) (uint32, uint32, error) {
+	manifest, err := s.loadDbtFile(dbtFile)
+	if err != nil {
+		return 0, 0, fmt.Errorf("load file %s: %w", dbtFile, err)
+	}
+
+	grants, filters, masks, err := s.loadAccessProvidersFromManifest(manifest)
+	if err != nil {
+		return 0, 0, fmt.Errorf("load access providers from manifest: %w", err)
+	}
+
+	grantIds, filterIds, maskIds, apsToRemove, err := s.loadExistingAps(ctx, grants, filters, masks)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return s.createAndUpdateAccessProviders(ctx, grants, grantIds, masks, maskIds, filters, filterIds, apsToRemove)
+}
+
+func (s *DbtService) createAndUpdateAccessProviders(ctx context.Context, grants map[string]*sdkTypes.AccessProviderInput, grantIds map[string]string, masks map[string]*sdkTypes.AccessProviderInput, maskIds map[string]string, filters map[string]*sdkTypes.AccessProviderInput, filterIds map[string]string, apsToRemove set.Set[string]) (uint32, uint32, error) {
+	numberOfChanges := len(grants) + len(masks) + len(filters) + len(apsToRemove)
+	var successUpdate uint32
+	var failed uint32
+
+	logChannel := make(chan bool) // channel will be true if ap is updated successfully.
+
+	createOrUpdateAp := func(name string, apInput *sdkTypes.AccessProviderInput, apIds map[string]string) (err error) {
+		defer func() {
+			if err != nil {
+				logChannel <- false
+			} else {
+				logChannel <- true
+			}
+		}()
+
+		if id, found := apIds[name]; found {
+			s.logger.Debug(fmt.Sprintf("update access provider %q (%q)", name, id))
+
+			_, updateErr := s.accessProviderClient.UpdateAccessProvider(ctx, id, *apInput, services.WithAccessProviderOverrideLocks())
+			if updateErr != nil {
+				return fmt.Errorf("update access provider %q (%q): %w", name, id, updateErr)
+			}
+		} else {
+			s.logger.Debug(fmt.Sprintf("create access provider %q", name))
+
+			_, createErr := s.accessProviderClient.CreateAccessProvider(ctx, *apInput)
+			if createErr != nil {
+				return fmt.Errorf("create access provider %q: %w", name, createErr)
+			}
+		}
+
+		return nil
+	}
+
+	var logWg = sync.WaitGroup{}
+	logWg.Add(1)
+
+	go func() {
+		defer logWg.Done()
+
+		for apUpdate := range logChannel {
+			if apUpdate {
+				successUpdate++
+			} else {
+				failed++
+			}
+
+			s.logger.Info(fmt.Sprintf("updated %d of %d access providers. %d successful, %d failures", successUpdate+failed, numberOfChanges, successUpdate, failed))
+		}
+	}()
+
+	var wg = multierror.Group{}
+
+	for key := range grants {
+		grant := grants[key]
+		name := key
+
+		wg.Go(func() error {
```

Comment:
I'm always worried when seeing this. When there is a lot to do, we would bombard our API a lot here. Even potentially run into DDOS protection issues (when really a lot :) )
Can we limit this with a workerpool?